azurerm/internal/acceptance/data.go

```diff
@@ -0,0 +1,71 @@
+package acceptance
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
+	"github.com/terraform-providers/terraform-provider-azuread/azuread/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/features"
+)
+
+type TestData struct {
+	// Location is a set of Azure Regions which should be used for this Test
+	Location Regions
+
+	// RandomString is a random integer which is unique to this test case
+	RandomInteger int
+
+	// RandomString is a random 5 character string is unique to this test case
+	RandomString string
+
+	// ResourceName is the fully qualified resource name, comprising of the
+	// resource type and then the resource label
+	// e.g. `azurerm_resource_group.test`
+	ResourceName string
+
+	// Environment is a struct containing Details about the Azure Environment
+	// that we're running against
+	Environment azure.Environment
+
+	// EnvironmentName is the name of the Azure Environment where we're running
+	EnvironmentName string
+
+	// resourceType is the Terraform Resource Type - `azurerm_resource_group`
+	resourceType string
+
+	// resourceLabel is the local used for the resource - generally "test""
+	resourceLabel string
+}
+
+// BuildTestData generates some test data for the given resource
+func BuildTestData(t *testing.T, resourceType string, resourceLabel string) TestData {
+	env, err := Environment()
+	if err != nil {
+		t.Fatalf("Error retrieving Environment: %+v", err)
+	}
+
+	testData := TestData{
+		RandomInteger:   tf.AccRandTimeInt(),
+		RandomString:    acctest.RandString(5),
+		ResourceName:    fmt.Sprintf("%s.%s", resourceType, resourceLabel),
+		Environment:     *env,
+		EnvironmentName: EnvironmentName(),
+
+		resourceType:  resourceType,
+		resourceLabel: resourceLabel,
+	}
+
+	if features.UseDynamicTestLocations() {
+		testData.Location = SupportedLocations()
```

Comment:
how will this handel tests that have to run in a specific location? or a location we don't have in any env var?

Might it make sense to pass in from a test "supported locations" and then if any of the ENV vars match its used? 

It seems `SupportedLocations` might be better phrased as `DesiredLocations` as support indicates it'll work

Reply:
for locations we don't have in an environment var (e.g. this is only available in this 1 region) we can continue hard-coding them as we do today.

we're only hard-coding to a single location in the tests AFAIK - so I think we can skip passing in `supported locations` for now, but if we need it that makes sense to me 👍 

in retrospect this can become a private function, so I guess we could split the difference and make this `availableLocations` since they're what the user running the test is saying is available to be used?